coderd/workspaceapps/appurl/appurl.go

```diff
@@ -83,6 +84,64 @@ func (a ApplicationURL) Path() string {
 	return fmt.Sprintf("/@%s/%s.%s/apps/%s", a.Username, a.WorkspaceName, a.AgentName, a.AppSlugOrPort)
 }
 
+func (a ApplicationURL) IsPort() bool {
+	// check if https port
+	if strings.HasSuffix(a.AppSlugOrPort, "s") {
+		trimmed := strings.TrimSuffix(a.AppSlugOrPort, "s")
+		_, err := strconv.ParseInt(trimmed, 10, 64)
+		//nolint:gosimple
+		if err != nil {
+			return false
+		}
+
+		return true
+	}
+
+	// check if port at all
+	_, err := strconv.ParseInt(a.AppSlugOrPort, 10, 64)
+	//nolint:gosimple
+	if err != nil {
+		return false
+	}
+
+	return true
+}
+
+func (a ApplicationURL) PortProtocol() string {
+	if strings.HasSuffix(a.AppSlugOrPort, "s") {
+		trimmed := strings.TrimSuffix(a.AppSlugOrPort, "s")
+		_, err := strconv.ParseInt(trimmed, 10, 64)
+		if err == nil {
+			return "https"
+		}
+	}
+
+	return "http"
+}
+
+func (a *ApplicationURL) ChangePortProtocol(target string) {
```

Comment:
If you have the function I suggested previously, this could be much simpler:
```
port, _, isPort := a.X()
if !isPort {
  return a
}
portStr := fmt.Sprint(port) // or probably strconv equivalent
if targetProtocol == "https" {
  portStr += "s"
}
return ApplicationURL{
  AppSlugOrPort: portStr,
  ...
}
```